source/openqasm/ast.py

```diff
@@ -0,0 +1,1010 @@
+
+
+@dataclass
+class Slice(IndexIdentifier):
+    """
+    Indexed identifier with a range
+    as subscript.
+
+    Example::
+
+        segments[1:2]
+        qubits[1:10:1]
+    """
+
+    name: str
+    range: RangeDefinition
+
+
+@dataclass
+class RangeDefinition(OpenNode):
+    """
+    Range definition.
+
+    Example::
+
+        [1:2]
+        [1:1:10]
+        [:]
+    """
+
+    start: Optional[Expression]
+    end: Optional[Expression]
+    step: Optional[Expression]
+
+
+@dataclass
+class Concatenation(IndexIdentifier):
+    """
+    Combination of two arrays.
+
+    Example::
+
+        segment1 || segment2
+        qubits[0:10] || qubits[15:20]
+    """
+
+    lhs: Union[IndexIdentifier, Identifier]
+    rhs: Union[IndexIdentifier, Identifier]
+
+
+@dataclass
+class CalibrationGrammarDeclaration(Statement):
+    """
+    Calibration grammar declaration
+
+    Example::
+
+        defcalgrammar "openpulse";
+    """
+
+    calibration_grammar: str
+
+
+@dataclass
+class CalibrationDefinition(Statement):
+    """
+    Calibration definition
+
+    Example::
+
+        defcal rz(angle[20] theta) $q {
+            shift_phase drive($q), -theta;
+        }
+    """
+
+    name: str
+    arguments: List[ClassicalArgument]
+    qubits: List[str]
+    return_type: Optional[ClassicalType]
+    body: str
+
+
+@dataclass
+class SubroutineDefinition(Statement):
+    """
+    Subroutine definition
+
+    Example::
+
+    def measure(qubit q) -> bit {
+        s q;
+        h q;
+        return measure q;
+    }
+    """
+
+    name: str
+    arguments: List[Union[ClassicalArgument, QuantumArgument]]
+    return_type: Optional[ClassicalType]
+    body: List[Statement]
+
+
+@dataclass
+class QuantumArgument(OpenNode):
+    """
+    Quantum argument in subroutine definition
+
+    Example::
+
+        qubit q
+
+        qubit[4] q
+
+    """
+
+    qubit: Qubit
+    designator: Optional[Expression]
+
+
+@dataclass
+class ReturnStatement(Statement):
+    """
+    Classical or quantum return statement
+
+    Example::
+
+        return measure q;
+
+        return a + b
+
+    """
+
+    expression: Optional[Union[Expression, QuantumMeasurement]]
+
+
+class ControlDirectiveStatement(Statement):
+    """
+    Base class for control directive statements
+    """
+
+
+class BreakStatement(ControlDirectiveStatement):
+    """
+    Break statement
+
+    Example::
+
+        break;
+
+    """
+
+
+class ContinueStatement(ControlDirectiveStatement):
+    """
+    Continue statement
+
+    Example::
+
+        continue;
+
+    """
+
+
+class EndStatement(ControlDirectiveStatement):
+    """
+    End statement
+
+    Example::
+
+        end;
+
+    """
+
+
+@dataclass
+class BranchingStatement(Statement):
+    """
+    Branch (if) statement
+
+    Example::
+
+        if(temp == 1) { ry(-pi / 2) scratch[0]; } else continue;
+
+    """
+
+    condition: Expression
+    if_block: List[Statement]
+    else_block: Optional[List[Statement]]
```

Comment:
Do we need to include the `Optional` in the `else_block`?  Obviously they are actually optional in the grammar, but storing it as an empty list conveys the same information (I think?), and doesn't imply that's there any behavioural difference between `else_block = None` and `else_block = []`.

Reply:
Good idea! Language like Kotlin intentional makes it hard to use null/None so one does not have to do null check. Changed.